docs/source/user_guide/writing-extractors.rst

```diff
@@ -0,0 +1,155 @@
+.. -*- mode: rst -*-
+.. vi: set ft=rst sts=4 ts=4 sw=4 et tw=79:
+
+.. _chap_writing-extractors:
+
+Writing custom extractors
+*************************
+
+An extractor, in MetaLad sense, is a class derived from one of the base extractor classes defined in ``datalad_metalad.extractors.base`` (``DatasetMetadataExtractor`` or ``FileMetadataExtractor``).
+It needs to implement several required methods, most notably ``extract()``.
+
+Example extractors can be found in MetaLad source code.
+
+.. TODO:: Add links to the example extractors
+
+Base class
+==========
+
+There are two primary types of extractors, dataset-level and file-level.
+
+Dataset-level extractors, by inheritance from the ``DatasetMetadataExtractor`` class, can access the dataset on which they operate as ``self.dataset``. Extractor functions may use it to call any dataset methods.
+
+File-level extractors, by inheritance from the ``FileMetadataExtractor``, can refer to the file on which it operates through ``self.file_info``.
+
+Required methods
+================
+
+``get_id()``
+------------
+
+This function should return a ``UUID`` object containing a UUID which is not used by another metalad-extractor or metalad-filter.
+Since it is meant to uniquely identify the extractor, its return value should be hardcoded, or generated in a deterministic way.
+
+A UUID can be generated in several ways, e.g. with Python (``import uuid; print(uuid.uuid4())``), or online generators
+(e.g. https://www.uuidgenerator.net/ or by searching 'UUID' in `DuckDuckGo <https://duckduckgo.com/>`_).
+
+While version 4 (random) UUIDs are probably sufficient, version 5 UUIDs (generated based on a given namespace and name) can also be used.
+For example, extractors in packages developed under the datalad organisation could use ``<extractor_name>.datalad.org``.
+Example generation with Python: ``uuid.uuid5(uuid.NAMESPACE_DNS, "bids_dataset.datalad.org")``.
+
+Example::
+
+  def get_id(self) -> UUID:
+      return UUID("d1e3f728-7868-4f33-a4a5-0f9f57dce342")
+
+``get_version()``
+-----------------
+
+This function should return a version string representing the extractor's version.
+
+The extractor version is meant to be included with the extracted metadata.
+It might be a version number, or a reference to a particular schema.
+While there are no requirements on the versioning scheme, it should be chosen carefully, and rules for version updating need to be considered.
+For example, when using `semantic versioning <https://semver.org/>`_ style, one might decide to guarantee that attribute type and name will never change within a major release, while new attributes may be added with a minor release.
```

Comment:
This ruleset borrowed from https://pypi.org/project/nbib/ which I used recently.